packages/pkl.experimental.structuredRead/structuredRead.pkl

```diff
@@ -0,0 +1,261 @@
+/// variables (if `inputScheme` is `env:`) into a structured and typed format.
+///
+/// When embeded in another application, this Pkl module can be used to provide a
+/// basic ability to override loaded configuration using CLI flags. The application
+/// will need to parse the passed flags, then provide them to Pkl via a custom
+/// [Resource Reader](https://pkl-lang.org/main/current/language-reference/index.html#extending-resource-readers),
+/// or by passing them in as external properties.
+function fill(target: Class|Module|Dynamic) =
+  let (targetClass =
+    if (!(target is Class))
+      target.getClass()
+    else
+      target
+  )
+    let (mappedInputs =
+      if (targetClass == Dynamic)
+        applyToDynamic(List(), target)
+      else
+        applyToClass(targetClass, List()))
+      let (failures = findFailures(mappedInputs))
+      if (failures.length > 0)
+        throwFailures(targetClass, failures)
+      else
+        if (targetClass != Dynamic)
+          deepToTyped.apply(targetClass, mappedInputs)
+        else
+          mappedInputs
+
+/// Sets the scheme used with `read()`. By default is `prop:` so values
+/// are read from external properties passed using the `-p` flag.
+///
+/// If set to `env:` this would result in values being read from
+/// environment variables.
+hidden inputScheme: String(matches(Regex(#"^[\d\w-]+:$"#)))
+
+/// Path seperator used to when computing the `read()` URI of a nested
+/// value. By default `.` which means that nested values are passed in
+/// using a dot seperated path as the key value.
+hidden pathSeperator: String
+
+/// Template used to convert a path of elements into a complete URI to
+/// be passed to `read?()`
+hidden pathTemplate: (List<String>) -> String = (path: List<String>) -> "\(inputScheme)\(path.join(pathSeperator))"
+
+/// Function used to read in values. Can be overwritten to change how
+/// values are read. Useful for testing
+hidden readFunc = (s) -> read?(s)
+
+function applyToClass(target: Class, path: List<String>): Mapping<String, Any?> =
+  let (reflected = reflect.Class(target))
+  new Mapping {
+    for (name, property in reflected.properties) {
+      when (isCoerceable(property)) {
+        when (isNullableProp(property)) {
+          [name] = let (potentialValue = readValueToTarget(path, name, property)) if (!(potentialValue is ReadFailure)) potentialValue else property.defaultValue
+        } else {
+          [name] = let (potentialValue = readValueToTarget(path, name, property)) if (!(potentialValue is ReadFailure)) potentialValue else property.defaultValue ?? potentialValue
+        }
+      }
+    }
+  }
+
+local function isCoerceable(prop: reflect.Property) =
+  let (_ = prop.type.getClass().simpleName)
+  (prop.type is reflect.DeclaredType) ||
+    (prop.type is reflect.NullableType) ||
+    (prop.type is reflect.TypeAlias) ||
+    (prop.type is reflect.UnknownType)
```

Comment:
I'm assuming the `let (_ =` is a remnant of debugging... Union types are a lovely thing. What about
```
typealias Coerceable =
  reflect.DeclaredType|reflect.NullableType|reflect.TypeAlias|reflect.UnknownType
```
and replacing `isCoerceable(foo)` with `foo is Coerceable`?